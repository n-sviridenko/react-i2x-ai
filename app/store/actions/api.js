import { createAction } from 'redux-actions';

export const REQUEST_FAIL = 'app/api/REQUEST_FAIL';

export const requestFail = createAction(REQUEST_FAIL);
