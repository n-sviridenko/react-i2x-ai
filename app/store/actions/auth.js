import { createAction } from 'redux-actions';
import { Map } from 'immutable';

export const LOGIN_REQUEST = 'app/auth/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'app/auth/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'app/auth/LOGIN_FAILURE';

export const loginRequest = createAction(LOGIN_REQUEST, (email, password) => Map({ email, password }));
export const loginSuccess = createAction(LOGIN_SUCCESS);
export const loginFailure = createAction(LOGIN_FAILURE);
