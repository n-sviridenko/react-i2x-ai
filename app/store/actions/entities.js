import { createAction } from 'redux-actions';

export const MERGE = 'app/entities/MERGE';

export const merge = createAction(MERGE);
