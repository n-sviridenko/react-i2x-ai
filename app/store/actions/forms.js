import { createAction } from 'redux-actions';

export const SUBMIT = 'app/forms/SUBMIT';

export const submit = createAction(SUBMIT, () => null, (options) => options);

export function buildSubmit(requestActionCreator, successActionType, failActionType) {
  return (values, dispatch) => new Promise((resolve, reject) => {
    const submitOptions = {
      requestAction: requestActionCreator(values),
      successActionType,
      failActionType,
      resolve,
      reject,
    };

    dispatch(submit(submitOptions));
  });
}
