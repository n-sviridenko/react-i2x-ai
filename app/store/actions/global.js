import { createAction } from 'redux-actions';

export const SET_LOCALE = 'app/global/SET_LOCALE';

export const setLocale = createAction(SET_LOCALE);
