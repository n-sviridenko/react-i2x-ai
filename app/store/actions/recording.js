import { createAction } from 'redux-actions';

export const LIST_REQUEST = 'app/recording/LIST_REQUEST';
export const LIST_SUCCESS = 'app/recording/LIST_SUCCESS';
export const LIST_FAIL = 'app/recording/LIST_FAIL';

export const listRequest = createAction(LIST_REQUEST);
export const listSuccess = createAction(LIST_SUCCESS);
export const listFail = createAction(LIST_FAIL);
