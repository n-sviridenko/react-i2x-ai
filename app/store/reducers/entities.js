import { fromJS } from 'immutable';
import { handleActions } from 'redux-actions';

import { MERGE } from 'store/actions/entities';

const initialState = fromJS({});

export default handleActions({
  [MERGE]: (state, { payload }) => state.mergeDeep(payload),
}, initialState);

export const getRoot = (state) => state.get('entities');
