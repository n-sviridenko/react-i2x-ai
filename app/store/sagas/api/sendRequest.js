import qs from 'qs';
import { fromJS } from 'immutable';
import { normalize } from 'normalizr';
import { call, put } from 'redux-saga/effects';

import request from 'utils/request';
import { merge } from 'store/actions/entities';
import { apiBaseUrl } from 'config';

export default function* sendRequest(method, endpoint, data = null, responseSchema = null, options = {}) {
  const headers = { ...options.headers };
  const finalMethod = method.toUpperCase();

  let url = `${apiBaseUrl}/${endpoint}`;
  let body;

  if (['GET', 'HEAD'].indexOf(finalMethod) !== -1) {
    const query = qs.stringify(data);

    if (query !== '') {
      url = `${url}?${query}`;
    }
  } else if (data instanceof FormData) {
    body = data;
  } else if (data) {
    headers['content-type'] = 'application/json';
    body = JSON.stringify(data);
  }

  const response = yield call(request, url, { ...options, method: finalMethod, headers, body });

  if (responseSchema === null) {
    return fromJS(response);
  }

  const { result, entities } = normalize(response, responseSchema);

  yield put(merge(fromJS(entities)));

  return fromJS(result);
}
