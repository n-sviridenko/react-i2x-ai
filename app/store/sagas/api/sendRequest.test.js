/* eslint-disable redux-saga/yield-effects */
import { call, put } from 'redux-saga/effects';
import { fromJS } from 'immutable';
import { schema } from 'normalizr';

import request from 'utils/request';
import { merge } from 'store/actions/entities';
import sendRequest from './sendRequest';

jest.mock('config', () => (
  {
    apiBaseUrl: 'http://api.io',
  }
));

describe('sendRequest', () => {
  it('should send "get" request', () => {
    const sendRequestTask = sendRequest('get', 'users', { foo: 'bar' });
    const callEffect = call(request, 'http://api.io/users?foo=bar', { method: 'GET', headers: {} });
    expect(sendRequestTask.next().value).toEqual(callEffect);
  });

  it('should send "post" request', () => {
    const sendRequestTask = sendRequest('post', 'users', { foo: 'bar' });
    const callEffect = call(
      request,
      'http://api.io/users',
      { method: 'POST', headers: { 'content-type': 'application/json' }, body: '{"foo":"bar"}' },
    );
    expect(sendRequestTask.next().value).toEqual(callEffect);
  });

  it('should send FormData', () => {
    const formData = new FormData();
    const sendRequestTask = sendRequest('post', 'users', formData);
    const callEffect = call(request, 'http://api.io/users', { method: 'POST', headers: {}, body: formData });
    expect(sendRequestTask.next().value).toEqual(callEffect);
  });

  it('should return immutable response', () => {
    const sendRequestTask = sendRequest('get', 'users');
    expect(sendRequestTask.next().value).toMatchSnapshot();
    const returnDescriptor = sendRequestTask.next({});
    expect(returnDescriptor.value).toEqual(fromJS({}));
    expect(returnDescriptor.done).toBe(true);
  });

  it('should return normalize response if schema is specified', () => {
    const responseSchema = new schema.Entity('user');
    const sendRequestTask = sendRequest('get', 'users/1', null, responseSchema);
    expect(sendRequestTask.next().value).toMatchSnapshot();
    const user = { id: 1, name: 'Sam' };
    const entities = fromJS({ user: { 1: user } });
    const mergeDescriptor = sendRequestTask.next(user);
    expect(mergeDescriptor.value).toEqual(put(merge(entities)));
    const returnDescriptor = sendRequestTask.next();
    expect(returnDescriptor.value).toBe(1);
    expect(returnDescriptor.done).toBe(true);
  });
});
