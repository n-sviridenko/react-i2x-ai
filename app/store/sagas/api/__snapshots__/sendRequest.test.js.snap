exports[`sendRequest should return immutable response 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://api.io/users",
      Object {
        "body": undefined,
        "headers": Object {},
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`sendRequest should return normalize response if schema is specified 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://api.io/users/1",
      Object {
        "body": undefined,
        "headers": Object {},
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
